src/timedate/timedated.c

```diff
@@ -48,19 +52,161 @@ static BUS_ERROR_MAP_ELF_REGISTER const sd_bus_error_map timedated_errors[] = {
         SD_BUS_ERROR_MAP_END
 };
 
+typedef struct UnitStatusInfo {
+        char *name;
+        char *load_state;
+        char *unit_file_state;
+        char *active_state;
+
+        LIST_FIELDS(struct UnitStatusInfo, units);
+} UnitStatusInfo;
+
 typedef struct Context {
         char *zone;
         bool local_rtc;
-        bool can_ntp;
-        bool use_ntp;
         Hashmap *polkit_registry;
+
+        LIST_HEAD(UnitStatusInfo, units);
 } Context;
 
+static void unit_status_info_clear(UnitStatusInfo *p) {
+        assert(p);
+
+        p->load_state = mfree(p->load_state);
+        p->unit_file_state = mfree(p->unit_file_state);
+        p->active_state = mfree(p->active_state);
+}
+
+static void unit_status_info_free(UnitStatusInfo *p) {
+        assert(p);
+
+        unit_status_info_clear(p);
+        free(p->name);
+        free(p);
+}
+
 static void context_free(Context *c) {
+        UnitStatusInfo *p;
+
         assert(c);
 
         free(c->zone);
         bus_verify_polkit_async_registry_free(c->polkit_registry);
+
+        while ((p = c->units)) {
+                LIST_REMOVE(units, c->units, p);
+                unit_status_info_free(p);
+        }
+}
+
+static int context_add_ntp_service(Context *c, const char *s) {
+        _cleanup_free_ char *name = NULL;
+        UnitStatusInfo *u;
+
+        if (endswith(s, ".service"))
+                name = strdup(s);
+        else
+                name = strjoin(s, ".service");
```

Comment:
hmm, should we really do such mangling? I don't think we should. I think it's entirely OK to expect folks to use fully qualified unit names here, it's low-level enough to expect that.

Moreover, if mangling really is desired then we should use the standard unit_name_mangle() function really, instead of something homegrown...